### PR TITLE
syntax/typedjson: expose shfmt's "typed JSON" as Go APIs

### LINKED ISSUE #885: Expose `ToJson` and `FromJson` methods
via https://github.com/mvdan/sh/issues/760#issuecomment-1120420888

`sh-syntax` (WASM) is about 2x faster than `mvdan-sh`

https://github.com/rx-ts/sh-syntax/blob/main/benchmark/benchmark.txt

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

Fixes #885.

### CODE REVIEW

**Comment by mvdan** [comment]
cc @JounQin @cilki

FYI @JounQin, the PR to make this work with tinygo will come after. Right now it doesn't work, but we are not far:

```
../syntax/typedjson/json.go:78:33: VisibleFields not declared by package reflect
../syntax/typedjson/json.go:89:21: StructOf not declared by package reflect
```

**Comment by JounQin** [comment]
It is so great to hear!

File: syntax/typedjson/json_test.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+			qt.Assert(t, err, qt.IsNil)
+			qt.Assert(t, sb.String(), qt.Equals, string(shellInput))
+
+			// Validate that emitting the JSON again produces the same result.
+			sb.Reset()
+			err = encOpts.Encode(sb, node2)
+			qt.Assert(t, err, qt.IsNil)
+			got = sb.String()
+			qt.Assert(t, got, qt.Equals, string(jsonInput))
+			t.Parallel()
```
Comment (riacataquian) [inline syntax/typedjson/json_test.go:83]: curious why is this in here rather than at the start?

**Review by riacataquian — APPROVED** [review]
The move to `typedjson` looks neat :)

File: syntax/typedjson/json_test.go (hunk 1 above)
Comment (mvdan) [reply syntax/typedjson/json_test.go:83]: yikes :) will fix.

**Comment by JounQin** [comment]
Hi, @mvdan , are we close to tinygo support? Is there still any blocking?

**Pull request merged**